### DISCUSSION TITLE
Error searching for documents: 57014 canceling statement due to statement timeout null

### DISCUSSION THREAD

**Discussion by simpel** [supabase/supabase#29767] [General]
Hi!

Whenever we try to do a similaritySearch on our vector DBs we get this error:

`Error searching for documents: 57014 canceling statement due to statement timeout null`

The db itself has something like 140 000 rows but the searches we're trying to do has really narrow metadata filters, for example checking for specific ID:s. The returned response really should be just a few documents. When we run our code in DBs with far less rows it works OK. 

1. Are there any settings we should look into? And how? Can we increase the timeout values for example?
2. Any thing special we should do with the DB? We just followed the LangChain examples here and then added loads of data basically: https://supabase.com/docs/guides/ai/langchain?queryGroups=database-method&database-method=sql

**Comment by youminkim** [comment]
Same issue here. We have never faced such issue last 2 years in Supabase but it started from last week. Pretty random and simple dropping table with 0 records returns timeout. We increase instance size and it works few days but it does not work today again

**Comment by iampopal** [comment]
I am also having the same issue. any update or a quick solution? Flutter Team?
@youminkim 
@simpel

**Comment by mattius459** [comment]
Same issue

**Comment by mbukeRepo** [comment]


**Reply by SelimEmre** [reply]
Hi @mbukeRepo, I'm having same issue. We are already using PRO plan. Which plan you upgraded?

**Comment by kon-rad** [comment]
same issue

**Comment by simpel** [comment]
Hi everyone! The ”solution” for us was to bump up the plan :(

**Comment by willy20040711** [comment]
@simpel @iampopal @SelimEmre 
I solved the problem by **set timeout for function match_documents**.
you can edit the function in the dashboard.
I use pro plan already.

`#variable_conflict use_column
begin
  return query
  select
    id,
    content,
    metadata,
    (embedding::text)::jsonb as embedding,
    1 - (documents.embedding <=> query_embedding) as similarity
  from documents
  where metadata @> filter
  order by documents.embedding <=> query_embedding
  limit match_count;
  set statement_timeout TO '5min'; 
end;`
below is the reference:
https://supabase.com/docs/guides/database/postgres/timeouts#function-level
https://supabase.com/docs/guides/ai/langchain?queryGroups=database-method&database-method=sql
https://supabase.com/docs/guides/troubleshooting/canceling-statement-due-to-statement-timeout-581wFv

**Comment by Admin368** [comment]
postgres=> alter role authenticator set statement_timeout = '10min'

helped me

Reference:
https://supabase.com/docs/guides/database/postgres/timeouts


## Vector Search Timeout Configuration (Critical for Large Datasets)

### Role-Level Timeout Setup
The following role timeout configuration is **required** for vector similarity searches with datasets over 25k records:

```sql
-- CRITICAL: Set timeout for authenticator role (this was the key fix)
ALTER ROLE authenticator SET statement_timeout = '2min';

-- Set timeouts for other roles
ALTER ROLE anon SET statement_timeout = '2min';
ALTER ROLE authenticated SET statement_timeout = '2min';
ALTER ROLE service_role SET statement_timeout = '5min';

-- Verify the configuration
SELECT rolname, rolconfig FROM pg_roles WHERE rolname IN ('authenticator', 'anon', 'authenticated', 'service_role');
```

### Incident Report: Vector Search Timeout (2025.08.18)

**Problem**: Supabase similarity search was giving "Error searching for documents: 57014 canceling statement due to statement timeout null" when vectorized dataset increased to over 25k records.

**Symptoms**:
- Searches consistently timing out at 8 seconds
- LangChain SupabaseVectorStore failing with statement timeout
- Function-level timeout increases not helping
- General PostgreSQL timeout increases not working

**Root Cause**: The `authenticator` role had a default timeout of 8 seconds that was overriding all other timeout settings.

**Solution**:
1. **Key Fix**: `ALTER ROLE authenticator SET statement_timeout = '2min';`
2. Updated other role timeouts for consistency
3. Set function-level timeout in `match_documents` function

**Default Supabase Role Timeouts**:
- `anon`: 3s
- `authenticated`: 8s  
- `service_role`: none (defaults to authenticator role's timeout)
- `authenticator`: 8s (this was the bottleneck!)
- `postgres`: none (capped by global timeout)

**References**:
- [Supabase Timeout Documentation](https://supabase.com/docs/guides/database/postgres/timeouts)
- [GitHub Discussion](https://github.com/orgs/supabase/discussions/29767)

**Testing**: Use `SELECT rolname, rolconfig FROM pg_roles;` to verify role configurations.
